Add support for loading lazy-loaded assemblies into DebugStore

captainsafia opened the pull request:
This PR introduces changes that allow assemblies that are lazily-loaded during the life-time of the application to be debugged.

* Adds `mono_wasm_add_lazy_load_files` function that can be invoked from JS by Blazor (see https://github.com/dotnet/aspnetcore/pull/25575 for example)
* Adds a special check on `Debugger.paused` event to evaluate `debugger` invocations triggered from the entrypoint above and passes them to the load infrastructure

I verified this manually by setting up Blazor's StandaloneApp with a lazily-loaded assembly (from a ProjectReference) and verifying that the sources show up under files and that a breakpoint in the sources is hit.

Commit:
Add support for loading lazy-loaded assemblies into DebugStore

thaystg commented:
Looks good to me, please add a test :)

radical commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -338,6 +339,11 @@ var MonoSupportLib = {
 			return exception_obj ;
 		},
 
+		mono_wasm_add_lazy_load_files: function(lazy_loaded_files) {
```
Can this get called more than once?

captainsafia replied:
Yes, it will be called whenever the user loads a set of assemblies using the `LazyLoaderService` in Blazor. Typically, this  will be called whenever a user navigates to a new page that loads assemblies lazily.

Any reservations about this?

radical replied:
Then should we clear `MONO.lazy_loaded_files` in `mono_wasm_get_lazy_loaded_files`, or when the existing list has been loaded?

captainsafia replied:
We can, although I'm not sure it is totally necessary. The flow would look something like this.

1. User navigates to page A.
2. Blazor app loads assemblies lazily.
3. Blazor app calls `mono_wasm_add_lazy_load_files`.
4. MONO.lazy_load_files is set to `lazy_loaded_files` parameter.
5. `mono_wasm_add_lazy_load_files` invokes `debugger;` to emit Debugger.paused event.
6. Proxy issues `mono_wasm_get_lazy_loaded_files` to get assemblies and loads them into the DebugStore.
7. User navigates to page B.
8. Blazor app loads another set of lazily-loaded assemblies.
7. Blazor app calls `mono_wasm_add_lazy_load_files` and the MONO.lazy_loaded_files list is overriden again.

So, I think it should be fine, but let me know if I am missing a subtlety here.

radical replied:
Silly question: When a user navigates to page B, it's like a new page was loaded, right? So, the runtime will be re-initialized, assemblies will be loaded etc? If so, then `mono_wasm_add_lazy_load_files` won't get called twice while in the same .. "session"?

captainsafia replied:
> When a user navigates to page B, it's like a new page was loaded, right? So, the runtime will be re-initialized, assemblies will be loaded etc? 

The navigation event occurs via a `history.pushState` event in the browser. The URL changes and Blazor reacts to that change and invokes the handlers that trigger the loading of assemblies. In this case, the runtime won't be re-initialized.

radical replied:
ok, so the state in `MONO`, would remain the same. Ok, so it might have stale list of files, but if we are only ever looking at it when paused due to this function being invoked, then we should be fine. Ok, understood.

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -498,6 +498,32 @@ async Task<bool> OnPause(SessionId sessionId, JObject args, CancellationToken to
+                    var res_val = loaded_assemblies.Value?["result"]?["value"];
+                    var lazy_loaded_files = res_val?.ToObject<string[]>();
+
+                    await
```
IIUC, this is the same as the code at https://github.com/dotnet/runtime/blob/dbacdc2acbb9173a07a09777e4383160c5cc9c80/src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs#L791 . We should extract this to a method.

captainsafia replied:
Sure! I can make this change.

radical commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -338,6 +339,11 @@ var MonoSupportLib = {
 			return exception_obj ;
 		},
 
+		mono_wasm_add_lazy_load_files: function(lazy_loaded_files) {
```
Can this be called while we are paused for some other reason, like a breakpoint? What does `debugger;` do in that case?

radical replied:
If it won't send a `debugger.Paused` again, then we won't hit the relevant code path in the proxy. If it does send it, then we wouldn't want to send `Debugger.resume` at the end of that.

captainsafia replied:
I tried this out on an E2E scenario by placing a breakpoint in the call to `LazyAssemblyLoader.LoadAssembliesAsync`. This is the C# method that makes the JS interop call to this method. I stepped into the code incrementally and it does dispatch the `Debugger.pause` event. The scripts are parsed and I am able to see them under the source view.

A `Debugger.resume` event does get dispatched but with the order of events it doesn't seem to have a break in the experience (e.g. it gets called but then a `StepInto` command is issued).

So practically, I'm not sure if there is a way for the user to end up in a position where a breakpoint is active while this method is being invoked, although I might be missing something in this trial I ran.

thaystg replied:
Is it possible to insert a breakpoint on cs file that will load the assembly dynamically? And try to step over? In this case would we break the step over of the user with our forced resume?

Commit:
Add end-to-end test and move check for top_func

captainsafia commented:
@thaystg @radical Pushed a commit here with some test-related changes.

- Created a new `lazy-debugger-test` project
  - Created new tests/debugger-test and tests/lazy-debugger-test directories for each project
  - Updated the Makefile to build the test projects
- Added `LoadSourcesFromFiles` shared by both initial load and lazy load entrypoints into the debugger
- Added test case to verify that script files are loaded into runtime

Part of the test case works. We are able to evaluate that the sources were properly loaded into the debugger.

However, I ran into some issues validating that a breakpoint can be set, specifically because the assembly is never loaded into the runtime. I tried to simulate loading the assembly into the runtime in two ways:

- Loading the assembly resource via a fetch request then passing the resulting array buffer to `mono_wasm_add_assembly`
- Loading the assembly resource via a fetch request then passing the byte array to a C# method that called `Assembly.Load` on the byte array

Neither of these approaches seemed to work. The first mimics what Blazor does when loading assemblies during initialization and the second mimics what Blazor does when loading assemblies lazily.

Let me know if you have any ideas on this.

[Here's a video of the E2E in Blazor](https://microsoft-my.sharepoint.com/:v:/p/safia/EffD7lTYqoZCsHy-8Fjvo64BRFEHZoAlpZrR6smq7wiVmw?e=WjW320), showing lazily loading a TestContentPackage assembly then debugging a component inside it.

thaystg commented on src/mono/wasm/Makefile:
```diff
@@ -146,20 +146,23 @@ run-tests-jsc-%:
+	cp $(TOP)/src/mono/wasm/debugger/tests/debugger-test/other.js $(TOP)/src/mono/wasm/debugger/tests/debugger-test/bin/Debug/publish
+	cp $(TOP)/src/mono/wasm/debugger/tests/debugger-test/runtime-debugger.js $(TOP)/src/mono/wasm/debugger/tests/debugger-test/bin/Debug/publish
+	cp $(TOP)/src/mono/wasm/debugger/tests/lazy-debugger-test/bin/Debug/publish/managed/lazy-debugger-test.dll $(TOP)/src/mono/wasm/debugger/tests/debugger-test/bin/Debug/publish
+	cp $(TOP)/src/mono/wasm/debugger/tests/lazy-debugger-test/bin/Debug/publish/managed/lazy-debugger-test.pdb $(TOP)/src/mono/wasm/debugger/tests/debugger-test/bin/Debug/publish
```
can you copy lazy-debugger-test.*

captainsafia replied:
Done!

thaystg commented on src/mono/wasm/debugger/DebuggerTestSuite/Tests.cs:
```diff
@@ -1575,7 +1575,7 @@ public async Task CreateGoodBreakpointAndHitAndRemoveAndDontHit()
                 var bp2 = await SetBreakpoint("dotnet://debugger-test.dll/debugger-test.cs", 12, 8);
                 var pause_location = await EvaluateAndCheck(
                     "window.setTimeout(function() { invoke_add(); invoke_add()}, 1);",
-                    "dotnet://debugger-test.dll/debugger-test.cs",  10, 8,
```
Unrelated change?

radical commented:
I have been looking at this, and was trying to get the assembly loaded. And I was able to do that by sending it as a base64 string to the test app, and then loading the assembly from C#.

But I feel that mono/wasm/debugger should automatically be updating the debug proxy, whenever the app loads an assembly at runtime. So, when Blazor loads the assembly, it shouldn't need to call into `library_mono.js`, or do anything else really.
Thoughts?

/cc @lewing @marek-safar @kg

Commit:
Fix up tests and address PR feedback

captainsafia commented:
Updated: the breakpointing part of E2E test is working now. Thanks for your help, @radical and @thaystg!

> But I feel that mono/wasm/debugger should automatically be updating the debug proxy, whenever the app loads an assembly at runtime. So, when Blazor loads the assembly, it shouldn't need to call into library_mono.js, or do anything else really.

That seems reasonably to me. Would be interested in seeing what API would be exposed for loading assemblies lazily. Blazor needs to do the actual fetching of the DLL via its resource loader (to do integrity checks, caching, etc). It would work if we had an `add_lazy_load_assembly` that took a set of byte streams for the DLLs and PDBs and then did the appropriate work under the hood.

radical commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:
```diff
@@ -2,7 +2,10 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.IO;
+using System.Runtime.Loader;
 using System.Threading.Tasks;
+
```
Um there are tests depending on exact line numbers in this file, so this will likely break them. My suggested code was just trying to get this working. Maybe you could use FQN for the new types being used here?

The test code itself should be fine. Also, could you run `make -C src/mono/wasm run-debugger-tests` locally? We don't have that enabled on the CI yet.

captainsafia replied:
Fixed.

captainsafia commented on src/mono/wasm/debugger/DebuggerTestSuite/Tests.cs:
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.IO;
```
My formatter got greedy here. I can revert if needed.

radical commented on src/mono/wasm/debugger/tests/debugger-test/debugger-driver.html:
```diff
@@ -73,6 +73,11 @@
 		}
+
+		function invoke_load_lazy_assembly() {
+			const files_to_load = ['http://localhost:9400/lazy-debugger-test.dll', 'http://localhost:9400/lazy-debugger-test.pdb'];
```
Can we build the url using `window.location`, or something?

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -99,6 +99,20 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, string meth
 
+                        if (top_func == "mono_wasm_add_lazy_load_files")
+                        {
+                            var loaded_assemblies = await SendMonoCommand(sessionId, MonoCommands.GetLazyLoadedFiles(), token);
```
Ensure that `loaded_assemblies.IsOk == true`.
nit: Maybe even rename that to res? and `res_val` to `loaded_assemblies`?

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -99,6 +99,20 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, string meth
+
+                            await SendCommand(sessionId, "Debugger.resume", new JObject(), token);
+                            return true;
+
```
Extra newline

radical commented on src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:
```diff
@@ -99,6 +99,20 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, string meth
+                        {
+                            var loaded_assemblies = await SendMonoCommand(sessionId, MonoCommands.GetLazyLoadedFiles(), token);
+                            var res_val = loaded_assemblies.Value?["result"]?["value"];
+                            var lazy_loaded_files = res_val?.ToObject<string[]>();
```
This can just resume+return, if there is nothing to be loaded.

radical commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs:
```diff
@@ -2,6 +2,8 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.IO;
+using System.Runtime.Loader;
```
(I could swear that I had added this comment already, but can't find it now!)
There are tests depending on exact line numbers in this file, so adding/removing them would break them. The test code itself should be fine though.

Also, could you please run `make -C src/mono/wasm run-debugger-tests` locally, because we don't have that enabled on the CI yet.

thaystg commented:
> /wasm/debugger should automatically be updating the debug proxy, whenever the app loads an assembly at runtime. So, when Blazor loads the assembly, it shouldn't need t

This was my first idea, I had discussed with Safia, mono runtime knows when we load a new assembly, and it doesn't matter how this is done, so mono runtime should be responsible to update debug proxy and not blazor.

captainsafia commented:
> This was my first idea, I had discussed with Safia, mono runtime knows when we load a new assembly, and it doesn't matter how this is done, so mono runtime should be responsible to update debug proxy and not blazor.

If I recall, that approach still had Blazor calling into the DebugProxy whenever an assembly was loaded, except over a WebSocket connection.

In any case, if this is something that needs to be modified within the runtime we should open an issue and move the discussion there.

radical commented:
> > This was my first idea, I had discussed with Safia, mono runtime knows when we load a new assembly, and it doesn't matter how this is done, so mono runtime should be responsible to update debug proxy and not blazor.
> 
> If I recall, that approach still had Blazor calling into the DebugProxy whenever an assembly was loaded, except over a WebSocket connection.
> 
> In any case, if this is something that needs to be modified within the runtime we should open an issue and move the discussion there.

I would imagine, that `mini-wasm-debugger.c` would have to somehow figure out that an assembly was loaded, and then it can  do something along the lines of what you have with `mono_wasm_add_lazy_load_files`. So, blazor can verify/cache the assembly, and when it does actually load it up, the proxy will get updated too. And Blazor shouldn't need to do anything else.
